### DISCUSSION TITLE
How to keep Canvas in sync going forward?

### DISCUSSION THREAD

**Discussion by cmichi** [paritytech/canvas#88] [General]
With the migration of the Canvas repo to a parachain, we originally had intended to get rid of the manual work of going through the commits in repos that affect us and apply them here as well. The idea was to "just" follow the [substrate-parachain-template](https://github.com/substrate-developer-hub/substrate-parachain-template) as close as possible ‒ automating as many steps of this as possible in the future. 

For our non-parachain [substrate-contracts-node](https://github.com/paritytech/substrate-contracts-node) this seems doable, we already have some helper scripts which display the `diff` to Substrate's [node-template](https://github.com/paritytech/substrate/tree/master/bin/node-template) and I think with a little more work we could automate this entirely.

With the parachain template there is an issue though: Rococo follows the `master` branches of Cumulus & Co., whereas the template follows the `release-vX.X.X` branches. So the template will always be behind what we need for running on Rococo.

I'm not sure what the best way forward is; maybe it just is to give up on this idea of easy synchronization and do what we did before: look through relevant commits and apply the changes manually. But before we stray further away from the idea of "easy synchronization" (e.g. https://github.com/paritytech/canvas/pull/87 will make this harder) we should reflect on it.

Maybe it would make sense to have a Rococo, Cumulus and Polkadot version of the parachain template? This would enable us to regularly clone that repository and automatically apply a defined set of changes on it. Could even be done in the spirit of a dependabot which creates a PR for Canvas and has somebody review it.

cc @riusricardo @NukeManDan @HCastano

**Comment by HCastano** [comment]
Substrate has the [`node-template`](https://github.com/paritytech/substrate/tree/master/bin/node-template) crate, which is used to keep up with the latest changes. There's also a few scripts ([1](https://github.com/paritytech/substrate/blob/master/.maintain/node-template-release.sh), [2](https://github.com/paritytech/substrate/tree/master/.maintain/node-template-release)) which are (maybe were?) used to update the Developer Hub's [`substrate-node-template`](https://github.com/substrate-developer-hub/substrate-node-template) repository. 

We might be able to use a similar approach by adding a `parachain-template` crate to Cumulus which is used to keep up with the latest changes, and then have some scripts which automate the process for updating the Develop Hub's `substrate-parachain-template` and `canvas`.

Right now we don't have a good reference within Cumulus since the `shell` parachain is too "dumb", and something like `statemine` is a little too opinionated. They're also using crates such as `parachains-common` which makes it hard to simply put the runtime into a new project.

@bkchr do you have any thoughts on how we can simply our life?

**Comment by riusricardo** [comment]
The thing with the **parachain-template** is that it isn't created from a script on the same way we do for the **node-template**.
The whole process requires cherry piking the changes from the Cumulus repo. Is a 100% manual process.
But I like the idea of moving the code of the parachain template (after some cleanups) to Cumulus as a way to maintain a working copy.
The reason why this hasn't been done is because Basti is the main workforce behind Cumulus and would add some maintenance burden.

**Reply by nuke-web3** [reply]
We already have a few production chains in cumulus so the core team must keep them building with changes as things are checked in ci/cd. The overhead to add one more that is a minimal template should be relatively small, and would help both reduce the overhead for the template for the devhub team, and ensure there is a template ready at a release (not after).

Frontier added a template generation script to make a stand alone temple from the main repo. This can be used in cumulus as well, IMHO it's the cleanest and best path forward.

Also tossing my hat in the ring to help maintain the templates, the parachain and frontier templates on devhub I presently am the primary on (with great support from the rest of the team of course!)

**Reply by bkchr** [reply]
> But I like the idea of moving the code of the parachain template (after some cleanups) to Cumulus as a way to maintain a working copy.

I don't see any problem with this

**Reply by cmichi** [reply]
I'm also in favor of this idea. @NukeManDan Could you then look into adding the template to Cumulus?

**Reply by HCastano** [reply]
I'm gonna snipe this from Dan and start working on a PR

**Comment by bkchr** [comment]
> With the parachain template there is an issue though: Rococo follows the `master` branches of Cumulus & Co., whereas the template follows the `release-vX.X.X` branches. So the template will always be behind what we need for running on Rococo.

That is not really true. While Rococo is using master, you still can connect as a parachain from one of the release branches. It is not that we break the networking or anything else, because we can not do this anymore. From time to time there are maybe some mismatches, but they are very unlikely. From a parachain perspective, using the release branch is completely fine.

**Reply by cmichi** [reply]
We had gone down this route (with the template based on v0.9.9), but ran into issues with incompatibility. IIRC we had to add changes to handle XCM v1, and there was also a new host function in Rococo that v0.9.9 wasn't compatible with.
